crates/extrinsics/src/instantiate.rs

```diff
@@ -377,7 +455,10 @@ impl Exec {
     }
 
     /// Dry run the instantiation before tx submission. Returns the gas required estimate.
-    async fn pre_submit_dry_run_gas_estimate(&self) -> Result<Weight> {
+    pub async fn pre_submit_dry_run_gas_estimate(
+        &self,
+        print_to_terminal: bool,
```

Comment:
Feels a bit odd for an API to have printing built in. Based on the code it seems possible to offload all printing to the caller.

Reply:
Thanks for the feedback! I agree that offloading printing to the caller could be a cleaner approach in some cases. However, in this specific scenario, there are a few reasons for the current design:

Printing statements are distributed throughout various sections of the function.
consider this part for example
```rust
match instantiate_result.result {
  Ok(_) => {
      if !self.output_json && print_to_terminal {
          super::print_gas_required_success(instantiate_result.gas_required);
  }
```
- The `instantiate_result.gas_required` value is printed within the function itself but is not returned as part of the function's result. Trying to mimic this behavior by moving the printing outside of the function would require replicating significant portions of the function's code in `cargo-contract` crate.
- Moving the printing logic outside of the API would require a significant amount of code changes throughout the `cargo-contract` crate. This could lead to a less maintainable codebase, and in some cases, it may even require exposing private internals, which is not ideal.
for example, we would need to expose functions like `print_gas_required_success` and `print_dry_running_status` which are now private.

The current approach with the printing parameter offers flexibility and convenience to users, especially when dealing with CLI tools. However, I think this makes for a good future improvement that might involve separating the printing logic from the API.

Reply:
>  moving the printing outside of the function would require replicating significant portions of the function's code in cargo-contract crate

So, this printing stuff is specific to cargo contract, and now this just shows why this code really doesn't belong into a generic contract extrinsic crate. Also the function is not very long, should be a rather small amount of work to move it into cargo contract. 

> Moving the printing logic outside of the API would require a significant amount of code changes throughout the cargo-contract crate

I don't see why a bunch of `print!` invocations requires significant code changes.

> we would need to expose functions like print_gas_required_success and print_dry_running_status

Can't this API here just expose a method to get the gas estimate and then anyone using it (for example cargo contract) can just build their own helpers which print stuff to the console around this if they want?

Reply:
Agreed
I've created dedicated helper functions within `cargo-contract` crate to handle both gas estimation along with printing. The API now exclusively offers `estimate_gas` methods for gas estimation, leaving the printing functionality to `cargo contract`'s own helper methods.